### PR TITLE
Convert `SamplerResult.quasi_dists` to list of `QuasiDistribution`

### LINKED ISSUE #362: Sampler returns SamplerResult with dict
**Describe the bug**

Sampler returns SamplerResult with a list of dict. However, it should be SamplerResult with a list of QuasiDistribution.
See https://github.com/Qiskit/qiskit-terra/blob/df54ae26d6125bee671ad27f4d412a92912aad1a/qiskit/primitives/sampler_result.py#L42 for detail.

====
This may cause bugs in the future. In particular, applications will be written assuming the type.

### PR DESCRIPTION
**Pull request by t-imamichi**
### Summary
Convert `SamplerResult.quasi_dists` to list of `QuasiDistribution` to be compatible with the [definition](https://github.com/Qiskit/qiskit-terra/blob/cf951202c77b49d28e7d23f43088b191638f282a/qiskit/primitives/sampler_result.py#L42)

### Details and comments
Fixes #362

### CODE REVIEW

**Comment by t-imamichi** [comment]
@ikkoham

**Review by rathishcholarajan — CHANGES_REQUESTED** [review]
@t-imamichi Looks good. Could you please add a release note since this is a breaking change?

**Comment by t-imamichi** [comment]
I added a reno. I see the discussion #362 is not concluded yet. It might be necessary to make this PR on hold for a while.

**Comment by t-imamichi** [comment]
@nonhermitian I updated this PR to convert mitigation overhead to stddev upper bound.

**Comment by nonhermitian** [comment]
@t-imamichi Looks good to me

**Comment by t-imamichi** [comment]
Thank you for reviewing it, @nonhermitian.

**Comment by t-imamichi** [comment]
I'm updating this PR to resolve conflicts.

**Comment by kevinsung** [comment]
Is this PR in conflict with https://github.com/Qiskit/qiskit-ibm-runtime/pull/511?

**Comment by t-imamichi** [comment]
Yes, it's conflicting.

**Comment by kt474** [comment]
> Yes, it's conflicting.

@t-imamichi is this ready to be merged?
If so, this can be merged first - #511 can wait (and needs some updates)

**Comment by t-imamichi** [comment]
I think it's ready. But, since @rathishcholarajan marked this on-hold, I want him to review it.

**Comment by t-imamichi** [comment]
Note that `QuasiDistribution` does not store the information of number of qubits and users need to set `binary_probabilities(num_bits)` by themselves. I have another PR to resolve this issue.
See https://github.com/Qiskit/qiskit-terra/pull/8464 for details.

File: qiskit_ibm_runtime/sampler.py
Comment (kevinsung) [inline qiskit_ibm_runtime/sampler.py:281]:
```suggestion
            if not shots:
                shots = float('inf')
            if overhead is None:
                overhead = 1.0
            # M3 mitigation overhead is gamma^2
            # https://github.com/Qiskit-Partners/mthree/blob/423d7e83a12491c59c9f58af46b75891bc622949/mthree/mitigation.py#L457
            #
            # QuasiDistribution stddev_upper_bound is gamma / sqrt(shots)
            # https://github.com/Qiskit/qiskit-terra/blob/ff267b5de8b83aef86e2c9ac6c7f918f58500505/qiskit/result/mitigation/local_readout_mitigator.py#L288
            stddev = sqrt(overhead / shots)
```

**Review by kevinsung — CHANGES_REQUESTED** [review]
Is this correct?

File: qiskit_ibm_runtime/sampler.py
Comment (t-imamichi) [reply qiskit_ibm_runtime/sampler.py:281]: Thank you for the suggestion. I can even simplify this approach.

Comment (t-imamichi) [reply qiskit_ibm_runtime/sampler.py]: See d52fa5a

File: qiskit_ibm_runtime/sampler.py
Comment (kevinsung) [inline qiskit_ibm_runtime/sampler.py:271]: This differs from my suggestion if `meta['shots'] == None` or `meta['readout_mitigation_overhead'] == None`. Is that possible? If so then the code below would raise an error.

Comment (t-imamichi) [reply qiskit_ibm_runtime/sampler.py:271]: These metadata are not designed to be set `None`. So, I think it's no problem.

**Comment by rathishcholarajan** [comment]
> I think it's ready. But, since @rathishcholarajan marked this on-hold, I want him to review it. I have just requested it to him.

@kt474 I put this `on hold` since it's a breaking change and better to release it along with the flexible session breaking changes. Will leave it up to you on when you want to merge this. PR looks good to me.

**Review by kt474 — APPROVED** [review]

**Pull request merged**
